### PR TITLE
Implement DirectUpload for Steps, Results and Inventories [SCI-3679]

### PR DESCRIPTION
**Pull request by okriuchykhin**
Jira ticket: [SCI-3679](https://biosistemika.atlassian.net/browse/SCI-3679)

### What was done
Implement DirectUpload for Steps, Results and Inventories

### CODE REVIEW

File: app/assets/javascripts/my_modules/results.js [hunk 1]
```diff
@@ -89,117 +65,113 @@
...
+      switch (resultTypeEnum) {
         case ResultTypeEnum.FILE:
-          _handleResultFileSubmit($form, ev);
+          handleResultFileSubmit($form, ev);
           break;
         case ResultTypeEnum.TABLE:
-          var $nameInput = $form.find('#result_name');
-          var nameValid = textValidator(ev, $nameInput, 0,
-            <%= Constants::NAME_MAX_LENGTH %>);
+          textValidator(ev, $form.find('#result_name'), 0, $form.data('name-max-length'));
```
Comment (aignatov-bio) [inline app/assets/javascripts/my_modules/results.js]: Maybe we can use this constant file - https://github.com/biosistemika/scinote-web/blob/activestorage_migration/app/assets/javascripts/sitewide/constants.js.erb

File: app/assets/javascripts/my_modules/results.js [hunk 2]
```diff
@@ -89,117 +65,113 @@
...
-            <%= Constants::NAME_MAX_LENGTH %>);
-          var $descrTextarea = $form.find("#result_text_attributes_textarea");
-          var $tinyMCEInput = TinyMCE.getContent();
-          textValidator(ev, $descrTextarea, 1, <%= Constants::RICH_TEXT_MAX_LENGTH %>, false, $tinyMCEInput);
-          break;
-        case ResultTypeEnum.COMMENT:
-          var $commentInput = $form.find('#comment_message');
-          var commentValid = textValidator(ev, $commentInput, 1,
-            <%= Constants::TEXT_MAX_LENGTH %>);
+          textValidator(ev, $form.find('#result_name'), 0, $form.data('name-max-length'));
```
Comment (aignatov-bio) [inline app/assets/javascripts/my_modules/results.js]: Same here

File: app/assets/javascripts/my_modules/results.js [hunk 3]
```diff
@@ -217,14 +189,40 @@
...
+      applyCollapseLinkCallBack();
+      applyCreateWopiFileCallback();
+
+      $(function() {
+        $('#results-collapse-btn').click(function() {
+          $('.result .panel-collapse').collapse('hide');
+          $(document).find('span.collapse-result-icon').each(function() {
+            $(this).addClass('fa-caret-square-down');
+            $(this).removeClass('fa-caret-square-up');
+          });
```
Comment (aignatov-bio) [inline app/assets/javascripts/my_modules/results.js]: You can make it without cycle:
`$(document).find('span.collapse-result-icon').addClass('fa-caret-square-down').removeClass('fa-caret-square-up')`

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 4]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+    TABLE.button(0).enable(false);
+  }
 
-  // Tells what action will execute by pressing on save button (update/create)
-  var saveAction = 'update';
-  var selectedRecord;
+  // Enable/disable edit button
+  function updateButtons() {
+    if (currentMode === 'viewMode') {
+      $('#addRepositoryRecord').removeClass('disabled');
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: Maybe we can combine objects by groups:
` $('#object1, #object2, ...').removeClass('disabled').prop('disabled', false)`

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 5]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+      }
+    }
+  }
+
+  // Helper functions
+  function listItemDropdown(options, currentValue, columnId) {
+    var html = '<select class="form-control selectpicker repository-dropdown" ';
+    html += 'data-abs-min-length="2" data-live-search="true" ';
+    html += 'data-container="body" column_id="' + columnId + '">';
+    html += '<option value="-1"></option>';
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: For multi-line strings we can use [Template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals), anyway we not support old IE browsers.

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 6]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+
+      // If checkbox is checked and row ID is not in list of selected row IDs
+      if (this.checked && index === -1) {
+        rowsSelected.push(rowId);
+      // Otherwise, if checkbox is not checked and row ID is in list of selected row IDs
+      } else if (!this.checked && index !== -1) {
+        rowsSelected.splice(index, 1);
+      }
+
+      if (this.checked) {
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: We can use here - `toggleClass`

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 7]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+
+  function appendInput(form, val, name) {
+    $(form).append(
+      $('<input>').attr('type', 'hidden').attr('name', name).val(val)
+    );
+  }
+
+  // Takes object and surrounds it with input
+  function changeToInputField(object, name, value) {
+    return "<div class='form-group'><input class='form-control' data-object='"
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: Also here use Template literals for better reading

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 8]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+        + object + "' name='" + name + "' value='" + value + "'></input></div>";
+  }
+
+  // Takes object and surrounds it with input
+  function changeToInputFileField(object, name, value) {
+    return "<div class='repository-input-file-field'>"
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: Also here use Template literals for better reading

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 9]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+    $('.modal-tooltip').off();
+    $('.modal-tooltip').hover(function() {
+      var $tooltip = $(this).find('.modal-tooltiptext');
+      var offsetLeft = $tooltip.offset().left;
+      var offsetTop = $tooltip.offset().top;
+      if ((offsetLeft + 200) > $(window).width()) {
+        offsetLeft -= 150;
+      }
+      $('body').append($tooltip);
+      $tooltip.css('background-color', '#d2d2d2');
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: I think better move it to CSS file and here simply add unique class.

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 10]
```diff
@@ -32,27 +50,429 @@ var RepositoryDatatable = (function(global) {
...
+    $('.dropdown-menu').parent()
+      .on('shown.bs.dropdown hidden.bs.dropdown', function() {
+        TABLE.columns.adjust();
+      });
+  }
+
+  // Clear all has-error tags
+  function clearAllErrors() {
+    // Remove any validation errors
+    $(selectedRecord).find('.has-error').each(function() {
```
Comment (aignatov-bio) [inline app/assets/javascripts/repositories/repository_datatable.js]: It also can be done without cycle:
`$(selectedRecord).find('.has-error').removeClass('has-error').find('span').remove()`

File: app/assets/javascripts/sitewide/drag_n_drop.js [hunk 11]
```diff
@@ -0,0 +1,630 @@
...
+      modal.on('hidden.bs.modal', function() {
+        modal.modal('hide').promise().done(function() {
+          modal.remove();
+        });
+        UPLOADED_IMAGE = {};
+      });
+    }
+
+    // Generate modal html and hook callbacks
+    function clipboardPasteModal() {
```
Comment (aignatov-bio) [inline app/assets/javascripts/sitewide/drag_n_drop.js]: Also we can use Template literals here.

File: app/assets/javascripts/sitewide/drag_n_drop.js [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+      } else {
+        $('.dnd-total-error').remove();
+        if (filerAndCheckFiles()) {
+          filesValid = true;
+          enableSubmitButton();
+        }
+      }
+    }
+
+    function uploadedAssetPreview(asset, i) {
```
Comment (aignatov-bio) [inline app/assets/javascripts/sitewide/drag_n_drop.js]: Also we can use Template literals here.

File: app/assets/javascripts/sitewide/drag_n_drop.js [hunk 13]
```diff
@@ -0,0 +1,630 @@
...
+        $(input).parent().find('.dnd-error').remove();
+        $(input).after("<p class='dnd-error'>" + I18n.t('general.text.length_too_long', { max_length: $(document.body).data('name-max-length') }) + '</p>');
+        isValid = false;
+      } else {
+        $(input).parent().find('.dnd-error').remove();
+        isValid = true;
+      }
+    }
+
+    function uploadedAssetPreview(asset, i) {
```
Comment (aignatov-bio) [inline app/assets/javascripts/sitewide/drag_n_drop.js]: Also we can use Template literals here.

File: app/assets/javascripts/repositories/repository_datatable.js (hunk 9 above)
Comment (okriuchykhin) [reply app/assets/javascripts/repositories/repository_datatable.js]: This is some hack for datatables

File: app/assets/javascripts/sitewide/tiny_mce.js [hunk 14]
```diff
@@ -20,80 +16,27 @@ var TinyMCE = (function() {
...
+
+
         tinyMCE.init({
           cache_suffix: '?v=4.9.3', // This suffix should be changed any time library is updated
           selector: selector,
           menubar: 'file edit view insert format',
-          toolbar: 'undo redo restoredraft | insert | styleselect | bold italic | alignleft aligncenter alignright alignjustify | bullist numlist outdent indent | link | forecolor backcolor | customimageuploader marvinjsplugin | codesample',
-          plugins: plugins,
+          toolbar: 'undo redo restoredraft | insert | styleselect | bold italic | alignleft aligncenter alignright alignjustify | bullist numlist outdent indent | link | forecolor backcolor | customimageuploader | codesample',
+          plugins: 'autosave autoresize customimageuploader link advlist codesample autolink lists charmap hr anchor searchreplace wordcount visualblocks visualchars insertdatetime nonbreaking save directionality paste textcolor placeholder colorpicker textpattern',
```
Comment (aignatov-bio) [inline app/assets/javascripts/sitewide/tiny_mce.js]: Can you add here marvinjsplugin and placeholder for tinyMCE

**Pull request merged**